### ISSUE TITLE
Train a Smooth Normalizing Flow with a score matching loss

### ISSUE THREAD

**Issue by EiffL** [astrodeepnet/sbi_experiments#13]

**Comment by Justinezgh** [comment]
So if I try to get the score with _loss=(score_hat - score)^2_,  _log_bound=4_ (meaning that max(a) approx 54) and _activation function=sin_, I have this : 
<img src="https://user-images.githubusercontent.com/72011736/147267274-0866af1d-e277-451f-b187-780c8653810a.png" width="300" height="250">

I think because of the behavior of both the forward/inverse ldj : 
<img src="https://user-images.githubusercontent.com/72011736/147236350-62519e56-7895-460d-91cc-35084fd96d2b.png" width="500" height="300">

So I tried with _log_bound=2_  (max(a) approx 7)  and _min_density_lower_bound=4e-1_ ( in order to have smoother forward/inverse ldj)
<img src="https://user-images.githubusercontent.com/72011736/147262856-4430e421-8136-4a8c-a6f0-b892531b2c9d.png" width="500" height="300">

but it still doesn't work. 

So I computed the grad(ildj) and grad(grad ildj) 
![image](https://user-images.githubusercontent.com/72011736/147268829-05cad503-2876-46fd-a475-66893a83c935.png)
and the grad(grad ildj) degenerates so the optimization seems to not be possible

**Comment by EiffL** [comment]
hummmmm interesting.... Which bijector is this? the cubic ramp?

**Comment by Justinezgh** [comment]
yup

**Comment by Justinezgh** [comment]
- x^5 
![image](https://user-images.githubusercontent.com/72011736/147276323-0af59ae9-fad2-4699-90b2-d06cfd2b12b4.png)

- x^7 
![image](https://user-images.githubusercontent.com/72011736/147276415-ab5eaaac-db0e-4344-90c1-ec0e75e6621f.png)

- x^9 
![image](https://user-images.githubusercontent.com/72011736/147276510-3af7db9f-4edc-44a0-a67e-02b54f4e0eb2.png)

**Comment by EiffL** [comment]
![image](https://user-images.githubusercontent.com/861591/147279013-c0571a5f-f291-40f8-8830-46c31da0ac9c.png)

not great.... I'm trying out their simple NonCompactAffineSigmoid to see what that does. It doesn't use a ramp at all, just a rescaling of a sigmoid function

**Comment by Justinezgh** [comment]
You mean this : 
<img width="338" alt="image" src="https://user-images.githubusercontent.com/72011736/147297040-06f21be4-4345-489a-a601-f10fe3c5382a.png">
?

**Comment by EiffL** [comment]
Humm yeah I think.

It seems to work decently, but our newton method to find the root fails from time to time, and it blows up during training... I'm trying to see what can be done to improve that....

**Comment by EiffL** [comment]
Ok, I've merged my code for this bijector above ^^^^^

Did you get a nicely working configuration?

**Comment by Justinezgh** [comment]
Yup so I used 3 CL and an exponential decay of the learning rate each 1000 updates.  [notebook](https://colab.research.google.com/drive/1WqpHHKEG_qHlA7JwB6W-EOweFVCaMR3F?usp=sharing)

Here are the results for 10 000 updates  (I saved the contour plot each 1000 updates to see the evolution) : 

- loss = nll 
<img src="https://user-images.githubusercontent.com/72011736/148528296-0261d3a0-cfc3-4587-a059-52a42976fece.png" width="1000" height="510">

- loss = score 
<img src="https://user-images.githubusercontent.com/72011736/148528380-d14d7361-374d-438d-95be-1b4e329281bc.png" width="1000" height="510">

- loss = score + nll 
<img src="https://user-images.githubusercontent.com/72011736/148528454-eff5014b-08b0-4ea0-aefb-0b563f764870.png" width="1000" height="510">

- loss = 2*score + nll 
<img src="https://user-images.githubusercontent.com/72011736/148528042-b03a55ff-db48-4d6a-8b27-f649e992a08a.png " width="1000" height="510">

- loss = score + 2*nll 
<img src="https://user-images.githubusercontent.com/72011736/148528531-26b349a2-aeae-4c32-b997-9befb58d7992.png" width="1000" height="510">

**Comment by EiffL** [comment]
nice :-D  ok so it looks like this model can be fitted to the target distribution quite well under an NLL. The score kinda works.... but is having a hard time with getting the relative amplitude of the two modes correctly.

In your combination experiment I guess you need to downweight the score loss a lot more, the value of the score loss is probably much larger than that of the NLL, by orders of mag, which means that most likely it dominates the 2 losses during training.

Because the NLL gives you a perfect result in this test, you won't be able to see an improvement with adding the score loss. So, I think you can try the following:
- Use a fixed batch of examples (e.g. 128)
- Train with NLL alone and observe that it overfits a lot
- (optional) Train with score loss alone and see how it looks
- Train different models with the NLL as the main loss but gradualy adding a small amount of score loss, and see if it starts improving things.

**Comment by EiffL** [comment]
I had 2 mins so I did a quick try with just 64 data points:

- Training with NLL alone:
![10](https://user-images.githubusercontent.com/861591/148550293-52fef0c7-c98a-4909-957c-4b65a299ff8b.png)

- Training with NLL + score/1000:
![10 (1)](https://user-images.githubusercontent.com/861591/148550354-3d0a334c-4d77-447b-944b-4e2ee4e97f39.png)

\o/
